src/drlx/trainer/dpo_trainer.py

```diff
@@ -0,0 +1,203 @@
+    ):
+        """
+        Get loss for training
+
+        :param chosen_batch_preds: Predictions for the ba
+        """
+        return self.sampler.compute_loss(
+            prompts=prompts, chosen_img=chosen_img, rejected_img=rejected_img,
+            denoiser=self.model, ref_denoiser=ref_denoiser, vae=self.vae,
+            device=self.accelerator.device,
+            method_config=self.config.method,
+            accelerator=self.accelerator
+        )
+    
+    @torch.no_grad()
+    def deterministic_sample(self, prompts):
+        """
+        Sample images deterministically. Utility for visualizing changes for fixed prompts through training.
+        """
+        gen = torch.Generator(device=self.pipe.device).manual_seed(self.config.train.seed)
+        self.pipe.unet = self.accelerator.unwrap_model(self.model).unet
+        return self.pipe(prompts, generator = gen).images
+
+    def train(self, pipeline):
+        """
+        Trains the model based on config parameters. Needs to be passed a prompt pipeline and reward function.
+
+        :param pipeline: Pipeline to draw tuples from with prompts
+        :type prompt_pipeline: DPOPipeline
+        """
+
+        # === SETUP ===
+        do_lora = self.config.model.lora_rank is not None
+
+        # Singular dataloader made to get a sample of prompts
+        # This sample batch is dependent on config seed so it can be same across runs
+        with scoped_seed(self.config.train.seed):
+            dataloader = self.accelerator.prepare(
+                pipeline.create_loader(batch_size = self.config.train.batch_size, shuffle = False)
+            )
+            sample_prompts = self.config.train.sample_prompts
+            if sample_prompts is None:
+                sample_prompts = []
+            if len(sample_prompts) < self.config.train.batch_size:
+                new_sample_prompts = next(iter(dataloader))["prompts"]
+                sample_prompts += new_sample_prompts
+                sample_prompts = sample_prompts[:self.config.train.batch_size]
+
+        # Now make main dataloader
+
+        assert isinstance(self.sampler, DPOSampler), "Error: Model Sampler for DPO training must be DPO sampler"
+
+        # Set the epoch count
+        epochs = self.config.train.num_epochs
+        if self.config.train.total_samples is not None:
+            epochs = int(self.config.train.total_samples // self.config.train.num_samples_per_epoch)
+
+        # Timer to measure time per 1k images (as metric)
+        timer = Timer()
+        def time_per_1k(n_samples : int):
+            total_time = timer.hit()
+            return total_time * 1000 / n_samples
+        last_batch_time = timer.hit()
+        
+        # Ref model
+        if not do_lora:
+            ref_model = deepcopy(self.accelerator.unwrap_model(self.model).unet)
+            ref_model.requires_grad = False
+            if self.config.method.ref_mem_strategy == "half": ref_model = ref_model.half()
```

Comment:
hmm is it a problem that the ref model is in half? does the original DPO do this?

Reply:
I'm not aware of how it would cause issues but I figured it was just a nice addition for memory saving purposes. It's not necessary if you match the original papers hyperparameters though.